### PR TITLE
Support SCTP port mapping (bump up API to v1.37)

### PR DESCRIPTION
**Pull request by ishidawataru**
(EDIT by @AkihiroSuda: add list for tracking vendor PRs)

Needs-vendoring:

- [x]  https://github.com/docker/libnetwork/pull/1825
- [x]  https://github.com/docker/swarmkit/pull/2298
- [x]  https://github.com/docker/go-connections/pull/41
- [x]  https://github.com/docker/cli/pull/278 (after this PR gets merged)

Vendored changes:

- libnetwork https://github.com/docker/libnetwork/compare/20dd462e0a0e883437a274bd61df4bc4de980830...ed2130d117c11c542327b4d5216a5db36770bc65
  - https://github.com/docker/libnetwork/pull/1825 Support SCTP port mapping
  - https://github.com/docker/libnetwork/pull/2070 Added OutboundNAT policy support for Windows
  - https://github.com/docker/libnetwork/pull/2071 Import the ssd tool in libnetwork (not relevant to code here)
- Go-connections: https://github.com/docker/go-connections/compare/98e7d807e5d804e4e42a98d74d1dd695321224ef...7beb39f0b969b075d1325fecb092faf27fd357b6
  - https://github.com/docker/go-connections/pull/41 Support parsing SCTP port mapping

- - - 
image

Signed-off-by: Wataru Ishida <ishida.wataru@lab.ntt.co.jp>

**- What I did**

This PR adds support for SCTP(Stream Control Transmission Protocol) port mapping.
SCTP is widely used in cellular networks as a transport protocol. One of the popular application of SCTP is [Diameter](https://en.wikipedia.org/wiki/Diameter_(protocol)). This PR enables running these applications inside docker containers.

Most of the work is done in libnetwork repo and I also opened a [PR](https://github.com/docker/libnetwork/pull/1825) for it. 

I'll shortly open a PR for docker/go-connections repo too.

vendor.conf is temporally filled with my forked repos.

Related issue: https://github.com/moby/moby/issues/9689

**- How I did it**

**- How to verify it**

use https://github.com/ishidawataru/cli/tree/sctp for the cli to try this out.

```
$ docker run -p 10000:20000/sctp mgoelzer/alpine-socat socat sctp-listen:20000 system:cat
```

```
$ docker run -it --net=host mgoelzer/alpine-socat socat stdin sctp4-connect:127.0.0.1:10000,bind=127.0.0.1
hello
hello
```

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Comment by AkihiroSuda** [comment]
```diff
diff --git a/hack/dockerfile/binaries-commits b/hack/dockerfile/binaries-commits
index 98344bc2f..ae6324541 100644
--- a/hack/dockerfile/binaries-commits
+++ b/hack/dockerfile/binaries-commits
@@ -6,9 +6,12 @@ TOMLV_COMMIT=9baf8a8a9f2ed20a8e54160840c492f937eeaf9a
 RUNC_COMMIT=2d41c047c83e09a6d61d464906feb2a2f3c52aa4
 CONTAINERD_COMMIT=3addd840653146c90a254301d6c3a663c7fd6429
 TINI_COMMIT=949e6facb77383876aeff8a6944dde66b3089574
-LIBNETWORK_COMMIT=7b2b1feb1de4817d522cc372af149ff48d25028e
+# TODO: remove LIBNETWORK_REPO when libnetwork#1825 gets merged to upstream
+LIBNETWORK_REPO=https://github.com/ishidawataru/libnetwork.git
+LIBNETWORK_COMMIT=c04444a9ef8bd272e9c7c971e6235f3a79e7b11c
 VNDR_COMMIT=c56e082291115e369f77601f9c071dd0b87c7120

 # CLI
-DOCKERCLI_REPO=https://github.com/docker/cli
-DOCKERCLI_COMMIT=3dfb8343b139d6342acfd9975d7f1068b5b1c3d3
+DOCKERCLI_REPO=https://github.com/ishidawataru/cli.git
+DOCKERCLI_COMMIT=49ee8ec5882240454c12e098ecc0b86fd12922c9
+
diff --git a/hack/dockerfile/install-binaries.sh b/hack/dockerfile/install-binaries.sh
index 370ec7ce4..9747478c7 100755
--- a/hack/dockerfile/install-binaries.sh
+++ b/hack/dockerfile/install-binaries.sh
@@ -40,7 +40,7 @@ install_containerd() {

 install_proxy() {
        echo "Install docker-proxy version $LIBNETWORK_COMMIT"
-       git clone https://github.com/docker/libnetwork.git "$GOPATH/src/github.com/docker/libnetwork"
+       git clone "$LIBNETWORK_REPO" "$GOPATH/src/github.com/docker/libnetwork"
        cd "$GOPATH/src/github.com/docker/libnetwork"
        git checkout -q "$LIBNETWORK_COMMIT"
        go build -ldflags="$PROXY_LDFLAGS" -o /usr/local/bin/docker-proxy github.com/docker/libnetwork/cmd/proxy
```

Please update hack/dockerfile temporarily for ease of testing and reviewing

**Comment by ishidawataru** [comment]
@AkihiroSuda done

File: hack/dockerfile/binaries-commits
Comment (AkihiroSuda) [inline hack/dockerfile/binaries-commits]: the commit seems gone. 72233db2ec641b322c2f49f81645ff5bf6f745b6?

Comment (ishidawataru) [inline hack/dockerfile/binaries-commits]: thanks, I pushed a fix.

**Comment by AkihiroSuda** [comment]
added code for swarm; please cherry-pick:
- https://github.com/AkihiroSuda/docker/commits/ishidawataru-sctp-1
- https://github.com/AkihiroSuda/libnetwork/commits/ishidawataru-sctp-1
- https://github.com/AkihiroSuda/swarmkit/commits/ishidawataru-sctp-1

**Comment by ishidawataru** [comment]
@AkihiroSuda Thanks a lot for your support! Cherry-picked your commit on my branches.
Also, updated the list of related PRs.

**Comment by AkihiroSuda** [comment]
CI failure is weird but seems real

**Comment by AkihiroSuda** [comment]
Test failure seems because this PR vendors CLI that seems recently got broken (https://github.com/docker/cli/issues/293)

@ishidawataru Please revert change to `DOCKERCLI_COMMIT` to see if CI still fails?
Sorry for going back and forth.

**Comment by ishidawataru** [comment]
@AkihiroSuda No problem. I reverted the commits.

**Comment by orbsfoc** [comment]
Any chance of this getting included?

**Comment by AkihiroSuda** [comment]
@orbsfoc https://github.com/docker/libnetwork/pull/1825 needs to be merged first. Review is welcome.

**Comment by Peter-eid** [comment]
@AkihiroSuda  @ishidawataru what is the status of this feature ?

**Comment by AkihiroSuda** [comment]
@Peter-eid

https://github.com/docker/libnetwork/pull/1825 needs to be reviewed and merged.

I guess not all libnetwork maintainers are familiar with SCTP, so inputs from SCTP    community would be helpful to move this forward.

**Comment by AkihiroSuda** [comment]
Rebased deps:
- docker/libnetwork#1825
- docker/swarmkit#2298
- docker/go-connections#41

Anyone PTAL?

**Comment by AkihiroSuda** [comment]
all deps have been merged now.

PTAL?

**Review by thaJeztah — APPROVED** [review]
Changes LGTM

@AkihiroSuda I opened https://github.com/moby/moby/pull/36335 to keep unrelated bumps out of this PR; we can rebase this PR after that's merged

@fcrisciani @nwoodmsft this vendor bump also brings in https://github.com/docker/libnetwork/pull/2070 - are local changes needed to use that? Does that change need updates in the documentation or CLI?

**Comment by thaJeztah** [comment]
https://github.com/moby/moby/pull/36335 was merged; can you rebase, @AkihiroSuda / @ishidawataru?

File: hack/dockerfile/binaries-commits
Comment (nwoodmsft) [reply hack/dockerfile/binaries-commits]: No local changes needed. From my perspective we are good to bring it in.

**Comment by AkihiroSuda** [comment]
rebased

**Comment by thaJeztah** [comment]
z failure doesn't look related; https://jenkins.dockerproject.org/job/Docker-PRs-s390x/8542/console

restarting

```
05:45:58 ----------------------------------------------------------------------
05:45:58 FAIL: check_test.go:366: DockerSwarmSuite.TearDownTest
05:45:58 
05:45:58 unmount of /tmp/docker-execroot/d90a4b96ddada/netns failed: invalid argument
05:45:58 unmount of /tmp/docker-execroot/d90a4b96ddada/netns failed: no such file or directory
05:45:58 check_test.go:371:
05:45:58     d.Stop(c)
05:45:58 daemon/daemon.go:391:
05:45:58     t.Fatalf("Error while stopping the daemon %s : %v", d.id, err)
05:45:58 ... Error: Error while stopping the daemon d29a6fde43de3 : exit status 130
05:45:58 
05:45:58 
05:45:58 ----------------------------------------------------------------------
05:45:58 PANIC: docker_api_swarm_test.go:461: DockerSwarmSuite.TestAPISwarmManagerRestore
05:45:58 
05:45:58 [d90a4b96ddada] waiting for daemon to start
05:45:58 [d90a4b96ddada] daemon started
05:45:58 
05:45:58 [d90a4b96ddada] exiting daemon
05:45:58 [d90a4b96ddada] waiting for daemon to start
05:45:58 [d90a4b96ddada] daemon started
05:45:58 
05:45:58 [d29a6fde43de3] waiting for daemon to start
05:45:58 [d29a6fde43de3] daemon started
05:45:58 
05:45:58 [d29a6fde43de3] exiting daemon
05:45:58 [d29a6fde43de3] waiting for daemon to start
05:45:58 [d29a6fde43de3] daemon started
05:45:58 
05:45:58 [dd4cd730508ff] waiting for daemon to start
05:45:58 [dd4cd730508ff] daemon started
05:45:58 
05:45:58 [dd4cd730508ff] exiting daemon
05:45:58 [dd4cd730508ff] waiting for daemon to start
05:45:58 [dd4cd730508ff] daemon started
05:45:58 
05:45:58 [dd4cd730508ff] waiting for daemon to start
05:45:58 [dd4cd730508ff] daemon started
05:45:58 
05:45:58 [d90a4b96ddada] exiting daemon
05:45:58 [d29a6fde43de3] daemon started
05:45:58 Attempt #2: daemon is still running with pid 9141
05:45:58 Attempt #3: daemon is still running with pid 9141
05:45:58 Attempt #4: daemon is still running with pid 9141
05:45:58 [d29a6fde43de3] exiting daemon
05:45:58 ... Panic: Fixture has panicked (see related PANIC)
05:45:58 
```

**Comment by thaJeztah** [comment]
oh, does this require an API version bump? (At least a mention in the API history: https://github.com/moby/moby/blob/master/docs/api/version-history.md)

**Review by vdemeester — APPROVED** [review]
LGTM 🐸

**Comment by AkihiroSuda** [comment]
bumped up API to v1.37

**Review by cpuguy83 — APPROVED** [review]
LGTM

**Review by thaJeztah — APPROVED** [review]
LGTM

**Pull request merged**
